Library/Homebrew/linkage_checker.rb

```diff
@@ -0,0 +1,121 @@
+require "set"
+require "keg"
+require "formula"
+
+class LinkageChecker
+  attr_reader :keg, :formula
+  attr_reader :brewed_dylibs, :system_dylibs, :broken_dylibs, :variable_dylibs
+  attr_reader :undeclared_deps, :reverse_links
+
+  def initialize(keg, formula = nil)
+    @keg = keg
+    @formula = formula || resolve_formula(keg)
+    @brewed_dylibs = Hash.new { |h, k| h[k] = Set.new }
+    @system_dylibs = Set.new
+    @broken_dylibs = Set.new
+    @variable_dylibs = Set.new
+    @undeclared_deps = []
+    @reverse_links = Hash.new { |h, k| h[k] = Set.new }
+    check_dylibs
+  end
+
+  def check_dylibs
+    @keg.find do |file|
+      next if file.symlink? || file.directory?
+      next unless file.dylib? || file.mach_o_executable? || file.mach_o_bundle?
+      file.dynamically_linked_libraries.each do |dylib|
+        @reverse_links[dylib] << file
+        if dylib.start_with? "@"
+          @variable_dylibs << dylib
+        else
+          begin
+            owner = Keg.for Pathname.new(dylib)
+          rescue NotAKegError
+            @system_dylibs << dylib
+          rescue Errno::ENOENT
+            @broken_dylibs << dylib
+          else
+            @brewed_dylibs[owner.name] << dylib
+          end
+        end
+      end
+    end
+
+    @undeclared_deps = check_undeclared_deps if formula
+  end
+
+  def check_undeclared_deps
+      filter_out = proc do |dep|
+        dep.build? || (dep.optional? && !dep.option_names.any? { |n| formula.build.with?(n) })
+      end
+      declared_deps = formula.deps.reject { |dep| filter_out.call(dep) }.map(&:name) +
+                      formula.requirements.reject { |req| filter_out.call(req) }.map(&:default_formula).compact
+      undeclared_deps = @brewed_dylibs.keys - declared_deps.map { |dep| dep.split("/").last }
+      undeclared_deps -= [formula.name]
+      undeclared_deps
+  end
+
+  def display_normal_output
+    display_items "System libraries", @system_dylibs
+    display_items "Homebrew libraries", @brewed_dylibs
+    display_items "Variable-referenced libraries", @variable_dylibs
+    display_items "Missing libraries", @broken_dylibs
+    display_items "Possible undeclared dependencies", @undeclared_deps
+  end
+
+  def display_reverse_output
+    return if @reverse_links.empty?
+    sorted = @reverse_links.sort
+    sorted.each do |dylib, files|
+      puts dylib
+      files.each do |f|
+        unprefixed = f.to_s.strip_prefix "#{@keg}/"
+        puts "  #{unprefixed}"
+      end
+      puts unless dylib == sorted.last[0]
+    end
+  end
+
+  def display_test_output
+    display_items "Missing libraries", @broken_dylibs
+    puts "No broken dylib links" if @broken_dylibs.empty?
+    display_items "Possible undeclared dependencies", @undeclared_deps
+    puts "No undeclared dependencies" if @undeclared_deps.empty?
+  end
+
+  def broken_dylibs?
+    !@broken_dylibs.empty?
+  end
+
+  def undeclared_deps?
+    !@undeclared_deps.empty?
```

Comment:
I object strongly and I'm the one who insisted on `!empty?` instead of `any?`. While the latter is more convenient to type and is more succinct, they are not equivalent and have different semantics<sup>(1)</sup> and performance characteristics<sup>(2)</sup>. (I realize that `any?` is used a lot in Homebrew code where it should have been `empty?`, but I'd personally prefer not to proliferate the use of `any?` in new code.)

<sup>(1)</sup> `[nil].any?` vs. `![nil].empty?`
<sup>(2)</sup> Querying the size of the container (usually in constant time) vs. iterating over all members of a container and executing a block on every single element (linear time; though it might be doing short-circuit evaluation, but that's not obvious from the documentation).